ui/de-lib/src/main/java/org/iplantc/de/shared/AsyncProviderWrapper.java

```diff
@@ -0,0 +1,55 @@
+package org.iplantc.de.shared;
+
+import org.iplantc.de.client.events.EventBus;
+import org.iplantc.de.shared.events.UserLoggedOutEvent;
+
+import com.google.gwt.inject.client.AsyncProvider;
+import com.google.gwt.user.client.rpc.AsyncCallback;
+import com.google.inject.Inject;
+
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+
+/**
+ * @author aramsey
+ *
+ * AsyncProviderWrapper should be used in place of AsyncProvider in order to
+ * standardize how we handle failures during the callback.
+ *
+ * AsyncProvider is used to perform code splitting in GWT.  When the code splitting download
+ * fails, like in the case where the user is deauthenticated prior to downloading the code split,
+ * DeCasAuthenticationEntryPoint correctly responds with a 401, but the client needs to handle
+ * the redirect to the landing page properly.
+ *
+ * When the failure happens, note that it's not an AuthenticationException or StatusCodeException,
+ * it's some sort of AsyncFragmentLoader$HttpDownloadFailure that I haven't found a way to explicitly
+ * check for.  If a method is found, you can probably switch from AsyncCallback to AsyncCallbackWrapper
+ * so all failures (RPC, websockets, or code splits) are handled in AsyncCallbackWrapper.
+ *
+ * Also note: it is NOT possible to test out code splitting with SDM running.
+ */
+public class AsyncProviderWrapper<T> {
+
+    Logger LOG = Logger.getLogger(AsyncProviderWrapper.class.getName());
+    
+    @Inject AsyncProvider<T> provider;
+    public AsyncProviderWrapper() {
+
+    }
+
+    public void get(final AsyncCallback callback) {
+        provider.get(new AsyncCallback<T>() {
+            @Override
+            public void onFailure(Throwable caught) {
+                LOG.log(Level.SEVERE, caught.getMessage());
+                EventBus.getInstance().fireEvent(new UserLoggedOutEvent());
+            }
```

Comment:
It would be nice if there were some way to examine the error to determine if the error was caused by an invalid session or by something else. In this case, anything that causes an error will result in the user being redirected to the landing page, whether or not the user's session is still active. If this is anything like the RPC callbacks, however, I suspect that it's difficult to find definitive information in the exception to determine the cause of the error, though.